File: extractPfam.py
#Function to parse through given pfam files and return a 2D list with the
#protein sequence information

#Takes a pfamLabel (string with the pfam filename. i.e. PF04526). This
#method will then parse the file and return a 2D list corresponding to the
#file
#
#In the returned 2D list, each entry (sublist) corresponds to a single
#protein. It will have the proteins sequence (string form) in the 0th
#index, followed by the starting location of the pfam itself in the 1st
#index and the ending location of the pfam in the 2nd index.
#
#See test main below for an example
#
#Let Aidan know if you have any questions
def parsePfam(pfamLabel):
    filename = "pFams/%s" % (pfamLabel)
    currSeqs = []
    with open(filename, "r") as currFamFile:
        for line in currFamFile:
            data = line.split()
            if (len(data) != 1):
                data[2] = int(data[2])
                data[3] = int(data[3])
                currSeqs.append(data)

    return currSeqs

def parsePfamGo(pfamLabel, goLabel, GoSeq):
    filename = "pFams/%s" % (pfamLabel)
    currSeqs = []
    with open(filename, "r") as currFamFile:
        for line in currFamFile:
            data = line.split()
            if (len(data) != 1):
                if (data[0] in GoSeq.keys() and goLabel in GoSeq[data[0]]):
                    data[2] = int(data[2])
                    data[3] = int(data[3])
                    currSeqs.append(data)

    return currSeqs



def parseGOLabels():
    GoSeq = {}
    with open("goa_human.gaf") as goFile:
        for line in goFile:
            data = line.split()
            if (data[0] == "UniProtKB"):
                if data[1] not in GoSeq.keys():
                    GoSeq[data[1]] = [data[3]]
                else:
                    GoSeq[data[1]].append(data[3])
        
    return GoSeq
    



#EXAMPLE!
def testMain():
    """
    currSeqs = parsePfam("PF04526")
    for seq in currSeqs:
        print "Protien name: ", seq[0]
        print "Entire protein sequence: ", seq[1]
        print "Pfam starting position: ", seq[2]
        print "Pfam ending position: ", seq[3], "\n"
        """

    GoSeq = parseGOLabels()
    currSeqs = parsePfamGo("PF03265", "GO:0004531", GoSeq)
    for seq in currSeqs:
        print "Protien name: ", seq[0]
        print "Entire protein sequence: ", seq[1]
        print "Pfam starting position: ", seq[2]
        print "Pfam ending position: ", seq[3], "\n"

testMain()

File: extractPfam.cpp
//Function to parse through given pfam files and return a 2D list with the
//protein sequence information

#include <string>
#include <iostream>
#include <vector>
#include <sstream>
#include <fstream>
#include <unordered_map>

using namespace std;

struct proteinSeq {
	string name;
	string seq;
	int start;
	int end;
};

typedef struct proteinSeq proteinSeq;

//split a large stringed sentence into a vector of words
vector<string> split(string& s, char delimiter) {
    vector<string> words;
    string currWord;
    basic_istringstream<char> wordStream(s);
    while (getline(wordStream, currWord, delimiter)) {
        words.push_back(currWord); 
    }
    return words;
}



/*
#Takes a pfamLabel (string with the pfam filename. i.e. PF04526). This
#method will then parse the file and return a 2D list corresponding to the
#file
#
#In the returned 2D list, each entry (sublist) corresponds to a single
#protein. It will have the proteins sequence (string form) in the 0th
#index, followed by the starting location of the pfam itself in the 1st
#index and the ending location of the pfam in the 2nd index.
#
#See test main below for an example
#
#Let Aidan know if you have any questions
*/
vector<proteinSeq> parsePfam(string pfamLabel) {
    string filename = "pFams/" + pfamLabel;
    vector<proteinSeq> currSeqs;
	ifstream proteinFile;
	proteinFile.open(filename);
	string currLine;
	while(getline(proteinFile, currLine)) {
		vector<string> data = split(currLine, ' ');
		if (data.size() != 1) {
			proteinSeq newSeq;
			newSeq.name = data.at(0);
			newSeq.seq = data.at(1);
			newSeq.start = stoi(data.at(2));
			newSeq.end = stoi(data.at(3));
			currSeqs.push_back(newSeq);
		}
	}

    return currSeqs;
}

vector<proteinSeq>  parsePfamGo(string pfamLabel, string goLabel, unordered_map<string, vector<string> > GoSeq) {
	(void) goLabel;
	(void) GoSeq;
    string filename = "pFams/" + pfamLabel;
	vector<proteinSeq> currSeqs;
	ifstream proteinFile;
	proteinFile.open(filename);
	string currLine;
	while(getline(proteinFile, currLine)) {
		vector<string> data = split(currLine, ' ');
		if (data.size() != 1) {
			//if (GoSeq.contains(data.at(0)) && find(GoSeq[data.at(0)].begin(), 
			//			GoSeq[data.at(0)].end(), goLabel) != GoSeq[data.at(0)].end()) {
			if (find(GoSeq[data.at(0)].begin(), 
						GoSeq[data.at(0)].end(), goLabel) != GoSeq[data.at(0)].end()) {
				proteinSeq newSeq;
				newSeq.name = data.at(0);
				newSeq.seq = data.at(1);
				newSeq.start = stoi(data.at(2));
				newSeq.end = stoi(data.at(3));
				currSeqs.push_back(newSeq);
			}
		}
	}

    return currSeqs;
}

unordered_map<string, vector<string> > parseGOLabels() {
    unordered_map<string, vector<string> > GoSeq;
	string filename = "goa_human.gaf";
	ifstream proteinFile;
	proteinFile.open(filename);
	string currLine;
	while(getline(proteinFile, currLine)) {
		vector<string> data = split(currLine, ' ');
        if (data.at(0) == "UniProtKB") {
			GoSeq[data.at(1)].push_back(data.at(2));
		}
	}
        
    return GoSeq;
}
    

//EXAMPLE!
int main() {
	unordered_map<string, vector<string> > GoSeq = parseGOLabels();

	/*
    vector<proteinSeq> currSeqs = parsePfam("PF04526");
    for (size_t i = 0; i < currSeqs.size(); i++) {
        cout << "Protein sequence name: " << currSeqs.at(i).name << "\n";
        cout << "Entire protein sequence: " << currSeqs.at(i).seq << "\n";
        cout << "Pfam starting Position: " << currSeqs.at(i).start << "\n";
        cout << "Pfam ending Position: " << currSeqs.at(i).end << "\n\n";
	}
	*/
	vector<proteinSeq> currSeqs = parsePfamGo("PF04526", "GO:0004531", GoSeq);
    for (size_t i = 0; i < currSeqs.size(); i++) {
        cout << "Protein sequence name: " << currSeqs.at(i).name << "\n";
        cout << "Entire protein sequence: " << currSeqs.at(i).seq << "\n";
        cout << "Pfam starting Position: " << currSeqs.at(i).start << "\n";
        cout << "Pfam ending Position: " << currSeqs.at(i).end << "\n\n";
	}
	return 0;
}
